geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java

```diff
@@ -88,204 +92,176 @@ public void tearDown() throws Exception {
-    regionpr.put("key1", "value1");
+    regionPR.put(key, value);
+    regionPR.put(key1, value1);
     TXManagerImpl mgr = (TXManagerImpl) cache.getCacheTransactionManager();
     mgr.begin();
-    assertEquals("value", regionpr.get("key"));
-    assertEquals("value1", regionpr.get("key1"));
+    assertEquals(value, regionPR.get(key));
+    assertEquals(value1, regionPR.get(key1));
     mgr.commit();
   }
 
 
   /**
-   * Test that two transactions with only read operations don't produce CommitConflictException
-   * This test fill some initial value on startup and after that create different threads with
-   * region#get operations
-   * Sync two different threads commit time through lock
+   * Test that two transactions with read and put operations produce CommitConflictException
    */
-  @Test(/* no exception expected */)
-  public void testDetectReadConflict()
-      throws CacheException, ExecutionException, InterruptedException {
+
+  @Test
+  public void readConflictsTransactionCanBlockWriteTransaction() {
     cache.close();
     createCache();
-    TXManagerImpl txMgrImpl = (TXManagerImpl) cache.getCacheTransactionManager();
-    // fill some initial key-value on start
-    txMgrImpl.begin();
-    this.region.put("key1", "value1"); // non-tx
-    txMgrImpl.commit();
-    ExecutorService executor = Executors.newFixedThreadPool(2);
-    for (int i = 0; i < 20; i++) {
-      final StampedLock lock = new StampedLock();
-      lock.asWriteLock().lock();
-      //
-      Future<?> future1 = executor.submit(() -> {
-        CacheTransactionManager txMgr2 = cache.getCacheTransactionManager();
-        txMgr2.begin();
-        region.get("key1"); // non-tx
-        lock.asReadLock().lock();
-        txMgr2.commit();
-      });
-      Future<?> future2 = executor.submit(() -> {
-        CacheTransactionManager txMgr3 = cache.getCacheTransactionManager();
-        txMgr3.begin();
-        region.get("key1"); // non-tx
-        lock.asReadLock().lock();
-        txMgr3.commit();
-      });
-      pause(100);
-      lock.asWriteLock().unlock();
-      future1.get();
-      future2.get();
+
+    region.put(key, value);
+    region.put(key1, value1);
+    TXManagerImpl txManager = (TXManagerImpl) cache.getCacheTransactionManager();
+    txManager.begin();
+    assertThat(region.get(key)).isSameAs(value);
+    region.put(key1, newValue1);
+    TXState txState =
+        (TXState) ((TXStateProxyImpl) TXManagerImpl.getCurrentTXState()).getRealDeal(null, null);
+    txState.setAfterReservation(() -> readTransactionAfterReservation());
+    Runnable task = () -> doPutOnReadKeyTransaction();
+    new Thread(task).start();
+    txManager.commit();
+    assertThat(region.get(key)).isSameAs(value);
+    assertThat(region.get(key1)).isSameAs(newValue1);
+  }
+
+  private void readTransactionAfterReservation() {
+    allowWriteTransactionToCommitLatch.countDown();
+    try {
+      allowReadTransactionToProceedLatch.await(10, TimeUnit.SECONDS);
```

Comment:
The intention behind GeodeAwaitility is to introduce a consistent and long timeout for use in Awaitility calls and other places such as this Latch await. The GeodeAwaitility is quite long but the test will still complete as quickly as possible. Also, the longer timeout will make the test robust in the face of a long stop-the-world GC pause or a CI environment with very slow CPU(s).:
```
allowReadTransactionToProceedLatch.await(GeodeAwaitility.getTimeout().getValueInMS(), TimeUnit.MILLISECONDS);